From a8d5f2f2bd3d549d68f189f541a04e74859ad114 Mon Sep 17 00:00:00 2001
From: Andrey Pronin <apronin@chromium.org>
Date: Wed, 28 Jun 2017 15:24:16 -0700
Subject: [PATCH] CHROMIUM: tpm: ignore failed selftest in probe

If a tpm is in failed self-test mode, ConstinueSelfTest
command results in TPM_FAILEDSELFTEST error, and probe
fails. The tpm device is not created, and that prevents
the OS from attempting any further recover operations
with the tpm. Instead, ignore the negative result of the
self test, and create the device - the chip is out there,
it's just in the failed state.

BUG=chromium:728130
TEST=put tpm in a failed selftest mode, boot, check that
     /dev/tpm0 is present, and GetCapability commands
     succeed.

Change-Id: I56b101850902d1983653bf29c8e0c662a5c7e8a0
Signed-off-by: Andrey Pronin <apronin@chromium.org>
Reviewed-on: https://chromium-review.googlesource.com/553422
Commit-Ready: Mattias Nissler <mnissler@chromium.org>
Reviewed-by: Mattias Nissler <mnissler@chromium.org>
---
 drivers/char/tpm/tpm-interface.c | 7 ++-----
 1 file changed, 2 insertions(+), 5 deletions(-)

diff --git a/drivers/char/tpm/tpm-interface.c b/drivers/char/tpm/tpm-interface.c
index d1deeca6cda6..7069fd33ccc5 100644
--- a/drivers/char/tpm/tpm-interface.c
+++ b/drivers/char/tpm/tpm-interface.c
@@ -1163,11 +1163,8 @@ int tpm1_auto_startup(struct tpm_chip *chip)
 	rc = tpm_get_timeouts(chip);
 	if (rc)
 		goto out;
-	rc = tpm_do_selftest(chip);
-	if (rc) {
-		dev_err(&chip->dev, "TPM self test failed\n");
-		goto out;
-	}
+	if (tpm_do_selftest(chip))
+		dev_err(&chip->dev, "TPM self test failed - ignoring\n");
 
 	return rc;
 out:
-- 
2.23.0.rc1.153.gdeed80330f-goog

